### PR TITLE
C# Telemetry Client Generation

### PR DESCRIPTION
**Pull request by awschristou**
This change automates the creation of the SDK Client for Toolkit Telemetry. Much of this resided in the Toolkit (as a manual process), and has been shifted here (with automation). Future work will automatically publish this package, which will then be integrated into the toolkit.

* added the [service definition file](/telemetry/service/service-model.json) for the telemetry service, which can be used as the central source for all toolkits
* A new solution and project `AwsToolkit.Telemetry.SDK` has been set up to represent the package containing the Telemetry SDK (service client)
* created an MsBuild script, which is responsible for generating the code (details below) that is placed into `AwsToolkit.Telemetry.SDK` and compiled
* added a CI check to verify that the `AwsToolkit.Telemetry.SDK` successfully generates and compiles

### How the Telemetry Client is generated ###
The Client generator is not a standalone package, so we build it and then run it.
1. Clone the [AWS .NET SDK](https://github.com/aws/aws-sdk-net) repo
1. Compile the SDK Generator from this repo
1. Run the SDK Generator using this repo's Telemetry Service definition to produce the Telemetry Client code
1. Copy portions of the generated code into the `AwsToolkit.Telemetry.SDK` project (the project is configured to be the artifact packaged for the Toolkit)

### How this change was verified ###

I took the Nuget Package that is produced when compiling locally, placed it in a temporary local NuGet source, and added it as a package reference in the Toolkit, replacing the existing Telemetry Client code. The Toolkit compiled and it was practically a drop-in replacement.

## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -11,3 +11,7 @@ build
 **/.vs
 **/bin
 **/obj
+
+# C# Telemetry Client Generator Related
+telemetry/csharp/buildtemp/**
```
Comment (hunterwerlla) [inline .gitignore:16]: Nit: I think `**/buildtemp` and `**/Generated` are fine to add to the gitignore as a whole

File: buildspec/csharpTests.yml [hunk 2]
```diff
@@ -22,6 +26,25 @@ phases:
...
             - dotnet test -c Release ./telemetry/csharp/AwsToolkit.Telemetry.sln --logger trx --results-directory ./.testresults
+
+            ############################################################################################################
+            # SDK Telemetry Client
+            ############################################################################################################
+
+            # Fixes for cloning aws net sdk and getting 'filename too long' errors
+            # LongPathsEnabled Reference:
+            # https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file#enable-long-paths-in-windows-10-version-1607-and-later
+            - '& Set-ItemProperty -Path "HKLM:\SYSTEM\CurrentControlSet\Control\FileSystem" -Name LongPathsEnabled -Value 1'
```
Comment (hunterwerlla) [inline buildspec/csharpTests.yml:37]: Can we set this at the top, it's not actually the build

File: telemetry/csharp/client-generator-inputs/telemetry-2017-07-25.customizations.json [hunk 3]
```diff
@@ -0,0 +1,3 @@
+{
+    "file-description": "This file is used by the SDK Client Generator. See telemetry/csharp/README.md for details"
+}
```
Comment (hunterwerlla) [inline telemetry/csharp/client-generator-inputs/telemetry-2017-07-25.customizations.json:3]: Adding a comment on ` telemetry/csharp/toolkit-telemetry.snk` is impossible, but do we need this file?

File: telemetry/service/service-model.json [hunk 4]
```diff
@@ -0,0 +1,205 @@
+{
```
Comment (hunterwerlla) [inline telemetry/service/service-model.json:1]: I'm excited for this, we can do vscode and jb next to no longer duplicate this file :D

File: telemetry/csharp/client-generator-inputs/telemetry-2017-07-25.customizations.json (hunk 3 above)
Comment (awschristou) [reply telemetry/csharp/client-generator-inputs/telemetry-2017-07-25.customizations.json:3]: Yup, it allows the compiled assembly to be "strongly typed", which is necessary when using it in the toolkit. Which reminds me, I will update the PR with how I verified the produced assembly.

**Pull request merged**